# 03_chunking_strategies: SEC annual 10-K report Chunking

This notebook evaluates and compares three chunking algorithms (Fixed Character Splitting, Recursive Character Splitting, and Semantic Chunking) on a financial SEC report context to analyze semantic boundary alignment.

### Chunking Formulations
1. **Fixed Character Splitting**: Hard split at a set character count limit. Often splits words or sentences in half, causing loss of contextual coherence.
2. **Recursive Character Splitting**: Evaluates a priority list of delimiters (typically `["

", "
", " ", ""]`) recursively to keep paragraphs, sentences, and words together in a single chunk.
3. **Semantic Chunking**: Measures cosine distances $d_i$ between successive sentence embeddings $S_i, S_{i+1}$:
   $$d_i = 1 - \text{CosineSimilarity}(\phi(S_i), \phi(S_{i+1}))$$
   Splits are created at indices where the distance exceeds a statistical percentile threshold $\theta$ (typically the 80th percentile of all distance margins):
   $$\theta = \text{Percentile}(\{d_1, d_2, \dots, d_{M-1}\}, 80)$$

In [1]:
from langchain_text_splitters import CharacterTextSplitter, RecursiveCharacterTextSplitter
from langchain_openai import OpenAIEmbeddings
import numpy as np

sec_10k_text = """
ITEM 1A. RISK FACTORS
Our business, financial condition, operating results, and cash flows can be affected by a number of factors.
1. Production and Supply Chain Disruption: We are dependent on battery cells and raw materials. Any shortage or price spike in lithium, nickel, or cobalt will degrade our margins.
2. Competition and Market Pressure: The automotive industry is highly competitive. Standard legacy OEMs are releasing alternative EV models, creating downward pressure on average selling prices.
3. Cybersecurity and Data Privacy: We collect large volumes of telemetry data. Any breach or leak of customer profiles will result in litigation and regulatory fines under GDPR and CCPA.
4. Regulatory Subsidies and Tax Credits: Our profitability relies on green carbon credits. Changes in federal policy that reduce these subsidies will negatively affect net income.
"""
print("Raw SEC text length:", len(sec_10k_text))

D:\Study\Prep\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Raw SEC text length: 874


In [2]:
# Character and Recursive Splitting
char_splitter = CharacterTextSplitter(chunk_size=120, chunk_overlap=20, separator="\n")
char_chunks = char_splitter.split_text(sec_10k_text)

rec_splitter = RecursiveCharacterTextSplitter(chunk_size=120, chunk_overlap=20, separators=["\n\n", "\n", " ", ""])
rec_chunks = rec_splitter.split_text(sec_10k_text)

print(f"Character Splitting: {len(char_chunks)} chunks.")
print(f"Recursive Splitting: {len(rec_chunks)} chunks.")

Created a chunk of size 179, which is longer than the specified 120


Created a chunk of size 194, which is longer than the specified 120


Created a chunk of size 186, which is longer than the specified 120


Character Splitting: 6 chunks.
Recursive Splitting: 10 chunks.


In [3]:
# Semantic Chunking
import os
from dotenv import load_dotenv
load_dotenv(dotenv_path=r"d:\\Study\\Prep\\.env")

embeddings = OpenAIEmbeddings()
sentences = [s.strip() for s in sec_10k_text.replace("\n", ". ").split(". ") if len(s.strip()) > 10]
sentence_embeds = embeddings.embed_documents(sentences)

distances = []
for i in range(len(sentences) - 1):
    vec1 = np.array(sentence_embeds[i])
    vec2 = np.array(sentence_embeds[i+1])
    dist = 1 - (np.dot(vec1, vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2)))
    distances.append(dist)

threshold = np.percentile(distances, 80)
semantic_chunks = []
current_chunk = [sentences[0]]

for i, dist in enumerate(distances):
    if dist > threshold:
        semantic_chunks.append(". ".join(current_chunk) + ".")
        current_chunk = [sentences[i+1]]
    else:
        current_chunk.append(sentences[i+1])
semantic_chunks.append(". ".join(current_chunk) + ".")

print(f"Semantic Chunking: {len(semantic_chunks)} chunks.")
print("Sample Semantic Chunk 1:\n", semantic_chunks[0])

Semantic Chunking: 3 chunks.
Sample Semantic Chunk 1:
 RISK FACTORS. Our business, financial condition, operating results, and cash flows can be affected by a number of factors.. Production and Supply Chain Disruption: We are dependent on battery cells and raw materials. Any shortage or price spike in lithium, nickel, or cobalt will degrade our margins.. Competition and Market Pressure: The automotive industry is highly competitive. Standard legacy OEMs are releasing alternative EV models, creating downward pressure on average selling prices..


### Output Explanation & Verification

#### Executed Results:
- **Raw Character Count**: `874` characters.
- **Fixed Character Splitting**: `6 chunks` (triggered warning flags: *'Created a chunk of size 179/194/186... longer than specified 120'* because no single `
` occurred within 120 character gaps).
- **Recursive Splitting**: `10 chunks` (avoided middle-word splitting by using hierarchy delimiters).
- **Semantic Chunking**: `3 chunks`.

#### Semantic Coherence Analysis:
- Semantic Chunking grouped the 874-character text into **exactly 3 chunks** based on topic similarity boundaries. 
- **Sample Semantic Chunk 1** successfully compiled the Risk Factors title, preamble, and point 1 and 2 together into a single contextual block:
  *'RISK FACTORS. Our business, financial condition, operating results, and cash flows can be affected by a number of factors.. Production and Supply Chain Disruption: We are dependent on battery cells and raw materials. Any shortage or price spike in lithium, nickel, or cobalt will degrade our margins.. Competition and Market Pressure: The automotive industry is highly competitive. Standard legacy OEMs are releasing alternative EV models, creating downward pressure on average selling prices..'*

This verifies that semantic chunking correctly groups text dynamically based on topical similarity shifts rather than static character counts.